gpmp2/kinematics/WorkspacePoseVelocityPrior.h

```diff
@@ -0,0 +1,125 @@
+ */
+class WorkspacePoseVelocityPrior
+    : public gtsam::NoiseModelFactor2<gtsam::Vector, gtsam::Vector> {
+ private:
+  // typedefs
+  typedef WorkspacePoseVelocityPrior This;
+  typedef gtsam::NoiseModelFactor2<gtsam::Vector, gtsam::Vector> Base;
+
+  // arm
+  Arm arm_;
+
+  // desired workspace pose
+  gtsam::Pose3 des_pose_;
+
+  // desired velocity (linear and angular)
+  gtsam::Vector6 des_vel_;
+
+ public:
+  /// shorthand for a smart pointer to a factor
+  typedef std::shared_ptr<This> shared_ptr;
+
+  /// Default constructor
+  WorkspacePoseVelocityPrior() {}
+
+  /**
+   * Constructor
+   * @param cost_model cost function covariance
+   * @param des_vel desired velocity (linear and angular)
+   */
+  WorkspacePoseVelocityPrior(gtsam::Key poseKey, gtsam::Key velKey,
+                             const gtsam::SharedNoiseModel& cost_model,
+                             const Arm& arm, const gtsam::Pose3& des_pose,
+                             const gtsam::Vector6& des_vel)
+      : Base(cost_model, poseKey, velKey),
+        arm_(arm),
+        des_pose_(des_pose),
+        des_vel_(des_vel) {}
+
+  ~WorkspacePoseVelocityPrior() {}
+
+  /// error function
+  gtsam::Vector evaluateError(
+      const gtsam::Vector& joint_conf, const gtsam::Vector& joint_rates,
+      gtsam::OptionalMatrixType H1 = nullptr,
+      gtsam::OptionalMatrixType H2 = nullptr) const override {
+    using namespace gtsam;
+
+    // fk
+    std::vector<Pose3> joint_pose;
+    std::vector<Vector6> joint_vel;
+    std::vector<Matrix> J_jpx_jp, J_jvx_jp, J_jvx_jv;
+    arm_.forwardKinematics(joint_conf, joint_rates, joint_pose, &joint_vel,
+                           &J_jpx_jp, &J_jvx_jp, &J_jvx_jv);
+
+    Matrix66 H_ep;
+    Vector pose_error = des_pose_.logmap(joint_pose[arm_.dof() - 1], {}, H_ep);
+    Vector vel_error = des_vel_ - joint_vel[arm_.dof() - 1];
+    if (H1) {
+      // Jacobian for the joint positions
+      *H1 = (Matrix(12, arm_.dof()) << H_ep * J_jpx_jp[arm_.dof() - 1],//
+                              - J_jvx_jp[arm_.dof() - 1]).finished();
+    }
+    if (H2) {
+      // Jacobian for the joint rates
+      *H2 = (Matrix(12, arm_.dof()) << Matrix::Zero(6, arm_.dof()), //
+                              -J_jvx_jv[arm_.dof() - 1]).finished();
+    }
+    Vector12 tot_result;
+    tot_result << pose_error, vel_error;
+    return tot_result;
+  }
+
+  /// @return a deep copy of this factor
+  gtsam::NonlinearFactor::shared_ptr clone() const override {
+    return std::static_pointer_cast<gtsam::NonlinearFactor>(
+        gtsam::NonlinearFactor::shared_ptr(new This(*this)));
+  }
+
+  /** print contents */
+  void print(const std::string& s = "",
+             const gtsam::KeyFormatter& keyFormatter =
+                 gtsam::DefaultKeyFormatter) const {
+    std::cout << s << "WorkspacePoseVelocityPrior :" << std::endl;
+    Base::print("", keyFormatter);
+    std::cout << "desired end-effector pose : ";
```

Comment:
Nitpick: capitalize the first letter so the output looks prettier?
Similarly for the below output of the end effector velocity.